lib/cli/watch-run.js

```diff
@@ -254,16 +261,21 @@ const createRerunner = (mocha, watcher, {beforeRun} = {}) => {
   // running.
   let runner = null;
 
+  const blastFullCache = !mocha.options.parallel;
```

Comment:
> An alternative is to always do a full blast, and then we can remove some conditional logic and simplify a bit.

Yeah this is a good question... I'd personally lean towards always blasting the full cache.

* It's odd for both contributors and users having different behaviors between modes
* Just in case, it'd be good to clamp down on this kind of issue whenever possible
 
...provided there isn't a noticeable performance hit. Have you observed any particular slowdowns trying this out?

We're still a swamped with picking the project up right now and other commitments, so we're not likely to be able to do a full performance dive anytime soon (#5027). If you can set that up to just get a quick baseline of whether this is bad, that'd be very helpful!

Note also that we wouldn't be able to merge this soon - we've got some more work to get through before we can start shipping `semver-major` changes.

Reply:
About benchmarking: I ran it three times with the conditional blasting, and three times with the latest commit always blasting. It looks negligible, in both scenarios milliseconds go up and down a bit, but are in the same range. Around mid July I can do a more decent benchmark, with potentially a few hundreds of dependencies, if you wish.

Results (only `test/integration/options/watch.spec.js`):

```
Conditionally blasting:

      ✔ reruns test when watched test file is touched (4017ms)
      ✔ reruns test when watched test file crashes (4012ms)
      ✔ reruns test when file matching --watch-files changes (4016ms)
      ✔ reruns test when file matching --watch-files is added (4017ms)
      ✔ reruns test when file matching --watch-files is removed (4015ms)
      ✔ does not rerun test when file not matching --watch-files is changed (4021ms)
      ✔ picks up new test files when they are added (4018ms)
      ✔ reruns test when file matching --extension is changed (4017ms)
      ✔ reruns when "rs\n" typed (4017ms)
      ✔ reruns test when file starting with . and matching --extension is changed (4021ms)
      ✔ ignores files in "node_modules" and ".git" by default (4023ms)
      ✔ ignores files matching --watch-ignore (4023ms)
      ✔ reloads test files when they change (4021ms)
      ✔ reloads test dependencies when they change (4024ms)
      ✔ reloads all required dependencies between runs (4023ms)
      ✔ reloads all required dependencies between runs when mutated from a hook (4021ms)
      ✔ respects --fgrep on re-runs (4028ms)
      ✔ should not leak event listeners (15039ms)
        ✔ reruns test when watched test file is touched (4019ms)
        ✔ reruns test when watched test file is crashed (4018ms)
        ✔ mochaHooks.beforeAll runs as expected (4021ms)
        ✔ mochaHooks.beforeEach runs as expected (4024ms)
        ✔ mochaHooks.afterAll runs as expected (4026ms)
        ✔ mochaHooks.afterEach runs as expected (4022ms)

      ✔ reruns test when watched test file is touched (4029ms)
      ✔ reruns test when watched test file crashes (4018ms)
      ✔ reruns test when file matching --watch-files changes (4029ms)
      ✔ reruns test when file matching --watch-files is added (4030ms)
      ✔ reruns test when file matching --watch-files is removed (4034ms)
      ✔ does not rerun test when file not matching --watch-files is changed (4022ms)
      ✔ picks up new test files when they are added (4029ms)
      ✔ reruns test when file matching --extension is changed (4016ms)
      ✔ reruns when "rs\n" typed (4013ms)
      ✔ reruns test when file starting with . and matching --extension is changed (4028ms)
      ✔ ignores files in "node_modules" and ".git" by default (4028ms)
      ✔ ignores files matching --watch-ignore (4019ms)
      ✔ reloads test files when they change (4033ms)
      ✔ reloads test dependencies when they change (4039ms)
      ✔ reloads all required dependencies between runs (4033ms)
      ✔ reloads all required dependencies between runs when mutated from a hook (4026ms)
      ✔ respects --fgrep on re-runs (4037ms)
      ✔ should not leak event listeners (15053ms)
        ✔ reruns test when watched test file is touched (4024ms)
        ✔ reruns test when watched test file is crashed (4030ms)
        ✔ mochaHooks.beforeAll runs as expected (4028ms)
        ✔ mochaHooks.beforeEach runs as expected (4032ms)
        ✔ mochaHooks.afterAll runs as expected (4024ms)
        ✔ mochaHooks.afterEach runs as expected (4042ms)

      ✔ reruns test when watched test file is touched (4042ms)
      ✔ reruns test when watched test file crashes (4012ms)
      ✔ reruns test when file matching --watch-files changes (4017ms)
      ✔ reruns test when file matching --watch-files is added (4036ms)
      ✔ reruns test when file matching --watch-files is removed (4024ms)
      ✔ does not rerun test when file not matching --watch-files is changed (4024ms)
      ✔ picks up new test files when they are added (4035ms)
      ✔ reruns test when file matching --extension is changed (4024ms)
      ✔ reruns when "rs\n" typed (4032ms)
      ✔ reruns test when file starting with . and matching --extension is changed (4025ms)
      ✔ ignores files in "node_modules" and ".git" by default (4025ms)
      ✔ ignores files matching --watch-ignore (4025ms)
      ✔ reloads test files when they change (4032ms)
      ✔ reloads test dependencies when they change (4032ms)
      ✔ reloads all required dependencies between runs (4036ms)
      ✔ reloads all required dependencies between runs when mutated from a hook (4032ms)
      ✔ respects --fgrep on re-runs (4049ms)
      ✔ should not leak event listeners (15054ms)
      when in parallel mode
        ✔ reruns test when watched test file is touched (4026ms)
        ✔ reruns test when watched test file is crashed (4018ms)
      with required hooks
        ✔ mochaHooks.beforeAll runs as expected (4024ms)
        ✔ mochaHooks.beforeEach runs as expected (4026ms)
        ✔ mochaHooks.afterAll runs as expected (4025ms)
        ✔ mochaHooks.afterEach runs as expected (4028ms)

Always blasting:

      ✔ reruns test when watched test file is touched (4026ms)
      ✔ reruns test when watched test file crashes (4022ms)
      ✔ reruns test when file matching --watch-files changes (4024ms)
      ✔ reruns test when file matching --watch-files is added (4026ms)
      ✔ reruns test when file matching --watch-files is removed (4025ms)
      ✔ does not rerun test when file not matching --watch-files is changed (4030ms)
      ✔ picks up new test files when they are added (4031ms)
      ✔ reruns test when file matching --extension is changed (4026ms)
      ✔ reruns when "rs\n" typed (4028ms)
      ✔ reruns test when file starting with . and matching --extension is changed (4024ms)
      ✔ ignores files in "node_modules" and ".git" by default (4023ms)
      ✔ ignores files matching --watch-ignore (4025ms)
      ✔ reloads test files when they change (4028ms)
      ✔ reloads test dependencies when they change (4025ms)
      ✔ reloads all required dependencies between runs (4025ms)
      ✔ reloads all required dependencies between runs when mutated from a hook (4039ms)
      ✔ respects --fgrep on re-runs (4024ms)
      ✔ should not leak event listeners (15055ms)
        ✔ reruns test when watched test file is touched (4012ms)
        ✔ reruns test when watched test file is crashed (4026ms)
        ✔ mochaHooks.beforeAll runs as expected (4026ms)
        ✔ mochaHooks.beforeEach runs as expected (4025ms)
        ✔ mochaHooks.afterAll runs as expected (4024ms)
        ✔ mochaHooks.afterEach runs as expected (4026ms)

      ✔ reruns test when watched test file is touched (4027ms)
      ✔ reruns test when watched test file crashes (4024ms)
      ✔ reruns test when file matching --watch-files changes (4025ms)
      ✔ reruns test when file matching --watch-files is added (4028ms)
      ✔ reruns test when file matching --watch-files is removed (4025ms)
      ✔ does not rerun test when file not matching --watch-files is changed (4023ms)
      ✔ picks up new test files when they are added (4031ms)
      ✔ reruns test when file matching --extension is changed (4025ms)
      ✔ reruns when "rs\n" typed (4027ms)
      ✔ reruns test when file starting with . and matching --extension is changed (4024ms)
      ✔ ignores files in "node_modules" and ".git" by default (4026ms)
      ✔ ignores files matching --watch-ignore (4024ms)
      ✔ reloads test files when they change (4025ms)
      ✔ reloads test dependencies when they change (4028ms)
      ✔ reloads all required dependencies between runs (4025ms)
      ✔ reloads all required dependencies between runs when mutated from a hook (4025ms)
      ✔ respects --fgrep on re-runs (4031ms)
      ✔ should not leak event listeners (15038ms)
        ✔ reruns test when watched test file is touched (4021ms)
        ✔ reruns test when watched test file is crashed (4026ms)
        ✔ mochaHooks.beforeAll runs as expected (4023ms)
        ✔ mochaHooks.beforeEach runs as expected (4026ms)
        ✔ mochaHooks.afterAll runs as expected (4025ms)
        ✔ mochaHooks.afterEach runs as expected (4023ms)

      ✔ reruns test when watched test file is touched (4018ms)
      ✔ reruns test when watched test file crashes (4018ms)
      ✔ reruns test when file matching --watch-files changes (4014ms)
      ✔ reruns test when file matching --watch-files is added (4015ms)
      ✔ reruns test when file matching --watch-files is removed (4016ms)
      ✔ does not rerun test when file not matching --watch-files is changed (4021ms)
      ✔ picks up new test files when they are added (4021ms)
      ✔ reruns test when file matching --extension is changed (4028ms)
      ✔ reruns when "rs\n" typed (4031ms)
      ✔ reruns test when file starting with . and matching --extension is changed (4025ms)
      ✔ ignores files in "node_modules" and ".git" by default (4025ms)
      ✔ ignores files matching --watch-ignore (4023ms)
      ✔ reloads test files when they change (4029ms)
      ✔ reloads test dependencies when they change (4025ms)
      ✔ reloads all required dependencies between runs (4017ms)
      ✔ reloads all required dependencies between runs when mutated from a hook (4028ms)
      ✔ respects --fgrep on re-runs (4038ms)
      ✔ should not leak event listeners (15049ms)
        ✔ reruns test when watched test file is touched (4023ms)
        ✔ reruns test when watched test file is crashed (4025ms)
        ✔ mochaHooks.beforeAll runs as expected (4032ms)
        ✔ mochaHooks.beforeEach runs as expected (4027ms)
        ✔ mochaHooks.afterAll runs as expected (4025ms)
        ✔ mochaHooks.afterEach runs as expected (4025ms)
```

Reply:
About full-blast always: I think it is the best approach, but I wanted to both play it safe and gather your feedback first. Removed the conditional logic and now always doing a full cache blast. Let me know if you see anything to improve

About when to release a new major: No problem, there's no hurry, and worst case scenario could just use my fork's branch.

Thanks for keeping the library alive and maintained!

Reply:
Great, sounds good to me! Full cache blasting is a 👍 from me. 🙂